api/python/t4/packages.py

```diff
@@ -841,3 +841,27 @@ def diff(self, other_pkg):
         added = list(sorted(other_entries))
         
         return added, modified, deleted
+
+    def map(self, f, include_directories=False):
+        if include_directories:
+            raise NotImplementedError
+
+        return [f(lk, entity) for lk, entity in self.walk()]
+
+    def filter(self, f, include_directories=False):
+        if include_directories:
+            raise NotImplementedError
+
+        return [(lk, entity) for lk, entity in self.walk() if f(lk, entity)]
+
+    def reduce(self, f, default=None, include_directories=False):
+        if include_directories:
+            raise NotImplementedError
+
+        entities = list(self.walk())
+        out = default if default is not None else entities[0]
+
+        for i in range(len(entities)):
```

Comment:
I think you're double counting entries[0] of there's no default.
Also you can probably use a straight enumeration instead of `range(len())`

Reply:
Good catch. New behavior:

        entities = list(self.walk())
        out = default if default is not None else entities[0]
        iters = entities if default is not None else entities[1:]

        for entry in iters:
            out = f(out, entry)